src/ir/module-utils.h

```diff
@@ -392,29 +392,29 @@ template<typename T> struct CallGraphPropertyAnalysis {
   }
 };
 
-// Helper function for collecting the type signatures used in a module
+// Helper function for collecting all the types that are declared in a module,
+// which means the HeapTypes (that are non-basic, that is, not eqref etc., which
+// do not need to be defined).
 //
-// Used when emitting or printing a module to give signatures canonical
-// indices. Signatures are sorted in order of decreasing frequency to minize the
+// Used when emitting or printing a module to give HeapTypes canonical
+// indices. HeapTypes are sorted in order of decreasing frequency to minize the
 // size of their collective encoding. Both a vector mapping indices to
-// signatures and a map mapping signatures to indices are produced.
-inline void
-collectSignatures(Module& wasm,
-                  std::vector<Signature>& signatures,
-                  std::unordered_map<Signature, Index>& sigIndices) {
-  struct Counts : public std::unordered_map<Signature, size_t> {
-    void note(Signature sig) { (*this)[sig]++; }
+// HeapTypes and a map mapping HeapTypes to indices are produced.
+inline void collectHeapTypes(Module& wasm,
+                             std::vector<HeapType>& types,
+                             std::unordered_map<HeapType, Index>& typeIndices) {
+  struct Counts : public std::unordered_map<HeapType, size_t> {
+    bool isRelevant(Type type) { return !type.isBasic() && type.isRef(); }
```

Comment:
What does the `&& type.isRef()` add?

Reply:
Hmm, isn't `funcref` both basic and reference? We need to ignore funcref here.

Reply:
Would't `!type.isBasic()` alone be enough to ignore `funcref`, though? Are you meaning to exclude RTT types here?

Reply:
Oh, I see what you mean now. Yes, ignoring all basic ones should be enough.

Reply:
Oh, no, that breaks all the tests, it lets something invalid through.

I don't think it's RTT though, since there are none?

Anyhow, l find it clearer as it is, plus it does seem good to disallow anything that is not a reference.

Reply:
Oh, maybe tuples? Anyway this is fine.